File: src/codes/duplicate-logic/py/rfc_strategy.py
from abc import ABC, abstractmethod

class ExportStrategy(ABC):
    @abstractmethod
    def export_data(self, data: str):
        pass

class PdfExportStrategy(ExportStrategy):
    def export_data(self, data: str):
        print(f"PDF出力: {data}")

class CsvExportStrategy(ExportStrategy):
    def export_data(self, data: str):
        print(f"CSV出力: {data}")

class XmlExportStrategy(ExportStrategy):
    def export_data(self, data: str):
        print(f"XML出力: {data}")

class Exporter:
    def __init__(self, strategy: ExportStrategy):
        self._strategy = strategy

    def set_strategy(self, strategy: ExportStrategy):
        self._strategy = strategy

    def export(self, data: str):
        print("開始ログ")
        self._strategy.export_data(data)
        print("完了ログ")

# 利用例
exporter = Exporter(PdfExportStrategy())
exporter.export("帳票データ")

exporter.set_strategy(CsvExportStrategy())
exporter.export("ユーザー一覧")

File: src/codes/duplicate-logic/ts/rfc_strategy.ts
// Strategy インターフェース
interface ExportStrategy {
  exportData(data: string): void;
}

// 各出力形式（ConcreteStrategy）
class PdfExportStrategy implements ExportStrategy {
  exportData(data: string): void {
    console.log(`PDF出力: ${data}`);
  }
}

class CsvExportStrategy implements ExportStrategy {
  exportData(data: string): void {
    console.log(`CSV出力: ${data}`);
  }
}

class XmlExportStrategy implements ExportStrategy {
  exportData(data: string): void {
    console.log(`XML出力: ${data}`);
  }
}

// コンテキスト
class Exporter {
  constructor(private strategy: ExportStrategy) {}

  setStrategy(strategy: ExportStrategy): void {
    this.strategy = strategy;
  }

  export(data: string): void {
    console.log("開始ログ");
    this.strategy.exportData(data);
    console.log("完了ログ");
  }
}

// 利用例
const exporter = new Exporter(new PdfExportStrategy());
exporter.export("帳票データ");

exporter.setStrategy(new CsvExportStrategy());
exporter.export("ユーザー一覧");
